crates/networking/p2p/sync.rs

```diff
@@ -0,0 +1,137 @@
+                // Discard the first header as we already have it
+                all_block_headers.extend_from_slice(&block_headers[1..]);
+                all_block_hashes.extend_from_slice(&block_hashes[1..]);
+
+                // Check if we already reached our sync head or if we need to fetch more blocks
+                if !block_hashes.contains(&sync_head) {
+                    // Update the request to fetch the next batch
+                    current_head = *block_hashes.last().unwrap();
+                } else {
+                    // No more headers to request
+                    break;
+                }
+            }
+        }
+        // We finished fetching all headers, now we can process them
+        // TODO: snap-sync: launch tasks to fetch blocks and state in parallel
+        // full-sync: Fetch all block bodies and execute them sequentially to build the state
+        match tokio::spawn(download_and_run_blocks(
+            all_block_hashes,
+            all_block_headers,
+            self.peers.clone(),
+            store.clone(),
+        ))
+        .await
+        {
+            Ok(Ok(())) => info!(
+                "Sync finished, time elapsed: {} secs",
+                start_time.elapsed().as_secs()
+            ),
+            Ok(Err(error)) => warn!(
+                "Sync failed due to {error}, time elapsed: {} secs ",
+                start_time.elapsed().as_secs()
+            ),
+            _ => warn!(
+                "Sync failed due to internal error, time elapsed: {} secs",
+                start_time.elapsed().as_secs()
+            ),
+        }
+    }
+
+    /// Creates a dummy SyncManager for tests where syncing is not needed
+    /// This should only be used in tests as it won't be able to connect to the p2p network
+    pub fn dummy() -> Self {
+        let dummy_peer_table = Arc::new(Mutex::new(KademliaTable::new(Default::default())));
+        Self {
+            snap_mode: false,
+            peers: dummy_peer_table,
+        }
+    }
+}
+
+/// Requests block bodies from peers via p2p, executes and stores them
+/// Returns an error if there was a problem while executing or validating the blocks
+async fn download_and_run_blocks(
+    mut block_hashes: Vec<BlockHash>,
+    mut block_headers: Vec<BlockHeader>,
+    peers: Arc<Mutex<KademliaTable>>,
+    store: Store,
+) -> Result<(), ChainError> {
+    loop {
+        let peer = peers.lock().await.get_peer_channels().await;
+        debug!("Requesting Block Bodies ");
+        if let Some(block_bodies) = peer.request_block_bodies(block_hashes.clone()).await {
+            let block_bodies_len = block_bodies.len();
+            debug!("Received {} Block Bodies", block_bodies_len);
+            // Execute and store blocks
+            for body in block_bodies.into_iter() {
+                // We already validated that there are no more block bodies than the ones requested
+                let header = block_headers.remove(0);
```

Comment:
It's not really clear to me what's happening here, why do we remove the first one? Does it always exist?

Reply:
We begin the cycle with a list of  headers (and its matching list of hashes) for all the blocks we will need to fetch and execute.
We send a request asking for the bodies of the list of hashes that we have.
The request may return less bodies than the ones we asked for, but never more (this is checked by the request method). So we will never remove a header that is not there.
When we execute the blocks we need the bodies and their headers and we also need to keep track of which blocks we already received bodies for so we don't request them again. By removing the headers and hashes from the list we can avoid cloning the headers and update our list of hashes for the next request.